source/extensions/dynamic_modules/dynamic_modules.h

```diff
@@ -0,0 +1,65 @@
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "absl/status/statusor.h"
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace DynamicModules {
+
+/**
+ * A class for loading and managing dynamic modules. This corresponds to a single dlopen handle.
+ * When the DynamicModule object is destroyed, the dlopen handle is closed.
+ *
+ * This class is supposed to be initialized once in the main thread and can be shared with other
+ * threads.
+ */
+class DynamicModule {
+public:
+  DynamicModule(void* handle) : handle_(handle) {}
+  ~DynamicModule();
+
+  /**
+   * Get a function pointer from the dynamic module with a specific type.
+   * @tparam T the function pointer type to cast the symbol to.
+   * @param symbol_ref the symbol to look up.
+   * @return the symbol if found, otherwise nullptr.
+   */
+  template <typename T> T getFunctionPointer(const absl::string_view symbol_ref) const {
+    static_assert(std::is_pointer<T>::value &&
+                      std::is_function<typename std::remove_pointer<T>::type>::value,
+                  "T must be a function pointer type");
+    return reinterpret_cast<T>(getSymbol(symbol_ref));
+  }
+
+private:
+  /**
+   * Get a symbol from the dynamic module.
+   * @param symbol_ref the symbol to look up.
+   * @return the symbol if found, otherwise nullptr.
+   */
+  void* getSymbol(const absl::string_view symbol_ref) const;
+
+  // The raw dlopen handle that can be used to look up symbols.
+  void* handle_ = nullptr;
```

Comment:
```suggestion
  void* handle_;
```

Always initialized in the constructor.